### PR TITLE
Add Span class

### PR DESCRIPTION
**Pull request by S-Dafarra**
This PR adds an implementation for Span (see discussion in https://github.com/robotology/idyntree/issues/430), importing the [GSL implementation](https://github.com/Microsoft/GSL/blob/master/include/gsl/span)

Also tests are added, but no documentation is available. 
It adds the namespace ``details`` in ``iDynTree::``. Maybe it is better to rename it as ``SpanImpl`` or similar.

cc @diegoferigo

### CODE REVIEW

File: cmake/iDynTreeOptions.cmake [hunk 1]
```diff
@@ -108,26 +108,26 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_BINDIR}"
...
+    set(CXX14_FLAGS "-std=c++14")
+  elseif(CXX_HAS_STD_CXX1Y)
+    set(CXX14_FLAGS "-std=c++1y")
   endif()
 
   set(CMAKE_CXX_EXTENSIONS OFF)
-  set(CMAKE_CXX_STANDARD 11)
-  set(CMAKE_CXX_STANDARD_REQUIRED 11)
+  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (traversaro) [inline cmake/iDynTreeOptions.cmake]: Thanks for fixing this. : )

File: src/core/include/iDynTree/Core/Span.h [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+//Most of this file has been taken from https://github.com/Microsoft/GSL/blob/master/include/gsl/span
+
+/*
+ * Copyright (C) 2015 Fondazione Istituto Italiano di Tecnologia
```
Comment (traversaro) [inline src/core/include/iDynTree/Core/Span.h]: 2018

**Comment by traversaro** [comment]
Obviously blocked by https://github.com/robotology/yarp/pull/1615 . 

Regarding the C++14, I just realized (sorry) the tricky part is that in this PR we are starting to use C++14 in the header, so we should not only make sure that the iDynTree's [compilation units](https://www.cs.auckland.ac.nz/references/unix/digital/AQTLTBTE/DOCU_015.HTM) are using the correct C++14 compilation flags, but that all compilation unit that depend include iDynTree headers are also compiled (at least) as C++14. 

This is typically implemented in CMake using `cmake-compile-features` ( https://cmake.org/cmake/help/v3.8/manual/cmake-compile-features.7.html ) that are exported as other usage requirements ( https://cmake.org/cmake/help/v3.0/manual/cmake-buildsystem.7.html#build-specification-and-usage-requirements ) and are propagated using `target_link_libraries`. See an example of their use in:
https://github.com/robotology/yarp/blob/245690a2bb2b371a81efe0da2f9dbbd9f34e53b2/src/libYARP_OS/CMakeLists.txt#L421 

My proposal for this PR is: 
* Bump the cmake minimum required to 3.5 (already announced in https://github.com/robotology/QA/issues/280)
* Remove all the usage of ` idyntree_enable_cxx11()` or ` idyntree_enable_cxx14()`, and also remove the definition of that macro 
* Add a a `target_compile_features(idyntree-core PUBLIC cxx_constexpr)` in core's `CMakeLists.txt` . The use of C++14 will be propagated to all other targets linking idyntree-core naturally. 
Let me know @S-Dafarra if you have problems with this path, if necessary I can do it in a separate PR.

**Comment by traversaro** [comment]
Agh, apparently there is a problem: the quite old gcc version available in Travis is creating problem in compiling the `cstdio` header in C++14 mode. : (

**Comment by traversaro** [comment]
Are we sure that we need C++14? I was initially confused by `gsl_constexpr14`, but `constexpr` is actually available since C++11. Note that this does not change the fact that we are starting to use C++11 in the headers, with the implication discussed in https://github.com/robotology/idyntree/pull/434#issuecomment-380165339 .

**Comment by S-Dafarra** [comment]
Unfortunately we need it. See for example https://github.com/robotology/idyntree/blob/85088a61deb831475263439c1d12a0bb0505b396/src/core/include/iDynTree/Core/Span.h#L133

**Comment by S-Dafarra** [comment]
Regarding 
> Add a a target_compile_features(idyntree-core PUBLIC cxx_constexpr) in core's CMakeLists.txt . The use of C++14 will be propagated to all other targets linking idyntree-core naturally.

I guess we need to change it given that ``cxx_constexpr`` is available since C++11 as you pointed out. In order to 
> the use of C++14 will be propagated to all other targets linking idyntree-core naturally

is it sufficient to add only one C++14 feature in ``target_compile_features`` or we need to add them all?

**Comment by francesco-romano** [comment]
1. Have you checked that you can transform the original MIT licence into the GPL2.1 of iDynTree?
2. Some documentation on the class to describe what is its purpose? (and how to use it?)

**Comment by S-Dafarra** [comment]
> Have you checked that you can transform the original MIT licence into the GPL2.1 of iDynTree?

I would leave this to @traversaro (https://github.com/robotology/idyntree/issues/430#issuecomment-379774054)

> Some documentation on the class to describe what is its purpose? (and how to use it?)

I would need it too :sob: For the time being we can refer also to https://en.cppreference.com/w/cpp/container/span

**Comment by traversaro** [comment]
Technically speaking, we are not relicensing the MIT part, and that is the reason why we are leaving the header as it is. We are including it, as permitted by the MIT license. When it is compiled as part of the LGPL2+ library, it automatically gets "licensed" as LGPL. This is my basic understanding, but as usual IANAL.

**Comment by traversaro** [comment]
The cool think is that the usage of the of the class is consistent with `std::vector`, i.e. you can access elements with `span[1]`, the size with `.size()` and the data with `data()`. 

To reduce the friction with iDynTree types, I think we need to sort out the conflict between `operator()` vs `operator[]` . I originally used  `operator()` because most of the SWIG languages that we target support the automatic wrapping of `operator()`, but nor `operator[]`. Note that I do not think that in Python/Matlab it make sense to manipulate wrapped `iDynTree::Span` objects, while I think the same object may be extremly useful for implementing custom [typemaps](http://www.swig.org/Doc3.0/Typemaps.html) (that however are implemented in C++).

**Comment by S-Dafarra** [comment]
> To reduce the friction with iDynTree types, I think we need to sort out the conflict between operator() vs operator[]

Maybe I did not get this right, but when calling the ``operator [ ]`` inside it does not call the correspondent container ``operator [ ]`` thus we can still have it defined in ``Span`` and not in the iDynTree types.

**Comment by traversaro** [comment]
> Maybe I did not get this right, but when calling the ``operator [ ]`` inside it does not call the correspondent container ``operator [ ]`` thus we can still have it defined in ``Span`` and not in the iDynTree types.

Yes, my comment was related to the consistency of the API.

**Comment by S-Dafarra** [comment]
> Yes, my comment was related to the consistency of the API.

Clear!
> Technically speaking, we are not relicensing the MIT part, and that is the reason why we are leaving the header as it is

Actually the header has been modified a bit . Does this matters?

**Comment by traversaro** [comment]
> Actually the header has been modified a bit . Does this matters? 

The MIT license permit to modify the content of the source code, so I do not think it is a big problem.

**Comment by traversaro** [comment]
* Due to all the C++14 usage, after this PR we should drop the support for Visual Studio 2013 . This is in line with what announced for YARP in https://github.com/robotology/QA/issues/280 , but I wonder if in the lab we still have to support VS2013 for some requirement of software such as FT drivers or something like that. @lucaTagliapietra @diegoferigo any idea on this? 

* From the AppVeyor output in https://ci.appveyor.com/project/robotology/idyntree/build/1.0.1473/job/g1akqbvc9ajlb4y5#L1031, we get:
~~~
CMake Error at src/core/CMakeLists.txt:132 (target_compile_features):
  target_compile_features The compiler feature "cxx_relaxed_constexpr" is not
  known to CXX compiler
  "MSVC"
  version 19.0.24215.1.
~~~
I guess the compiler feature "cxx_relaxed_constexpr"  is not supported at all in VS2015 . Given that GSL is supported on VS2015 (see https://github.com/Microsoft/GSL#supported-platforms), I guess this is a CMake issue, and we should probably just use a more "normal" C++14 feature, such as `cxx_attribute_deprecated` .

**Comment by traversaro** [comment]
Travis fixed in https://github.com/robotology/idyntree/pull/434/commits/9adeeefec3c27eaaaed15cf8aa44982cef8e716e .

**Comment by traversaro** [comment]
Rebased and squashed all the changes in https://github.com/robotology/idyntree/pull/434/commits/3dfb447bb80fb47076ed6e7b01166de85348cca7 .

**Comment by traversaro** [comment]
Merging as soon as AppVeyor is successful.

**Pull request merged**